### PR TITLE
Parallel isin

### PR DESCRIPTION
**Pull request by Rubtsowa**
| SDC1/SDC2 | SDC1/SDC4 | SDC1/SDC8 | SDC1/SDC16 | SDC1/SDC28 | SDC1/SDC56
-- | -- | -- | -- | -- | -- | --
Series.isin(master) | 1.001822826 | 0.999252002 | 0.997879899 | 0.999698454 | 0.999492456 | 1.00004082
Series.isin(use prange) | 0.295510625 | 0.297127618 | 0.303279895 | 0.303284178 | 0.325097053 | 0.243064832

How can it be improved?

### CODE REVIEW

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 1]
```diff
@@ -1990,7 +1990,13 @@ def hpat_pandas_series_isin(self, values):
     def hpat_pandas_series_isin_impl(self, values):
         # TODO: replace with below line when Numba supports np.isin in nopython mode
         # return pandas.Series(np.isin(self._data, values))
-        return pandas.Series(data=[(x in values) for x in self._data], index=self._index, name=self._name)
+
+        data_len = len(self._data)
+        result = numpy.array([False] * data_len)
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Didn't you try to create empty array of boolean type?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 2]
```diff
@@ -1990,7 +1990,13 @@ def hpat_pandas_series_isin(self, values):
...
+        for i in numba.prange(data_len):
+            result[i] = self._data[i] in values
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Didn't you try to convert `values` to set?

Comment (Rubtsowa) [reply sdc/datatypes/hpat_pandas_series_functions.py]: This slows down the execution of the function, even on lists with large repetitions.

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: @Rubtsowa 
no, it doesn't

name|nthreads|type|size|median|min|max|compile|boxing
----|--------|----|----|------|---|---|-------|------
Series.isin|1.0|Python|10000000.0|0.009159|0.009127|0.009223|NaN|NaN
 | | 1.0 | SDC|10000000.0|0.027854|0.027554|0.028204|0.104338|0.000232
||8.0|Python|10000000.0|0.009403|0.008639|0.012322|NaN|NaN
||8.0|SDC|10000000.0|0.006398|0.006361|0.007950|0.106473|0.00027


Moreover, I would say, it speeds everything up

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Without set:

name|nthreads|type|size|median|min|max|compile|boxing
----|--------|----|----|------|---|---|-------|------
Series.isin|1.0|Python|10000000.0|0.008859|0.008839|0.008965|NaN|NaN
||1.0|SDC|10000000.0|0.127839|0.127763|0.127877|0.095031|0.000235
||8.0|Python|10000000.0|0.009093|0.008691|0.012178|NaN|NaN
||8.0|SDC|10000000.0|0.324509|0.322576|0.454564|0.08109|0.000293

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Initial version (list comprehension):

name|nthreads|type|size|median|min|max|compile|boxing
----|--------|----|----|------|---|---|-------|------
Series.isin|1.0|Python|10000000.0|0.005723|0.005515|0.006462|NaN|NaN
||1.0|SDC|10000000.0|0.252864|0.232145|0.253373|0.066061|0.000244
||8.0|Python|10000000.0|0.005694|0.005566|0.006526|NaN|NaN
||8.0|SDC|10000000.0|0.232615|0.232179|0.232622|0.06533|0.000257

Comment (Rubtsowa) [reply sdc/datatypes/hpat_pandas_series_functions.py]: without set:
Ran 1 test in 0.505s

with set:
Ran 1 test in 1.676s

Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Is it time of execution of unit test or perf. test? Why do you use `python.unittest` time from command line?

**Comment by AlexanderKalistratov** [comment]
Looks like we need different implementation for string case

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 3]
```diff
@@ -1962,10 +1963,31 @@ def hpat_pandas_series_isin(self, values):
...
         ty_checker.raise_exc(values, 'set or list', 'values')
 
-    def hpat_pandas_series_isin_impl(self, values):
-        # TODO: replace with below line when Numba supports np.isin in nopython mode
-        # return pandas.Series(np.isin(self._data, values))
-        return pandas.Series(data=[(x in values) for x in self._data], index=self._index, name=self._name)
+    if isinstance(values[0], (types.UnicodeType, types.StringLiteral)):
+        def hpat_pandas_series_isin_impl(self, values):
+            # TODO: replace with below line when Numba supports np.isin in nopython mode
+            # return pandas.Series(np.isin(self._data, values))
```
Comment (kozlov-alexey) [inline sdc/datatypes/hpat_pandas_series_functions.py]: I think this comment is no longer needed and can be removed

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 4]
```diff
@@ -1962,10 +1963,31 @@ def hpat_pandas_series_isin(self, values):
     if not isinstance(values, (types.Set, types.List)):
         ty_checker.raise_exc(values, 'set or list', 'values')
 
-    def hpat_pandas_series_isin_impl(self, values):
-        # TODO: replace with below line when Numba supports np.isin in nopython mode
-        # return pandas.Series(np.isin(self._data, values))
-        return pandas.Series(data=[(x in values) for x in self._data], index=self._index, name=self._name)
+    if isinstance(values[0], (types.UnicodeType, types.StringLiteral)):
```
Comment (kozlov-alexey) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Hm, values[0] doesn't make sense for a Set (not at runtime, not at compile time), it's really strange that this works at all. Probably it should be:
values.dtype
```suggestion
    if isinstance(values.dtype, (types.UnicodeType, types.StringLiteral)):
```

**Comment by kozlov-alexey** [comment]
>  	SDC1/SDC2	SDC1/SDC4	SDC1/SDC8	SDC1/SDC16	SDC1/SDC28	SDC1/SDC56
> Series.isin(master)	1.001822826	0.999252002	0.997879899	0.999698454	0.999492456	1.00004082
> Series.isin(use prange)	0.295510625	0.297127618	0.303279895	0.303284178	0.325097053	0.243064832
> How can it be improved?

Can you please update this and attach here latest perf results?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 5]
```diff
@@ -1962,10 +1963,31 @@ def hpat_pandas_series_isin(self, values):
...
+        def hpat_pandas_series_isin_impl(self, values):
+            # TODO: replace with below line when Numba supports np.isin in nopython mode
+            # return pandas.Series(np.isin(self._data, values))
+
+            values = str_list_to_array(list(values))
+            values = set(values)
```
Comment (kozlov-alexey) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Have you measured how much time this part takes comparing to total execution?
And is there a perf test for this usecase already?

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.isin | 1 | Python | 10000000 | 0.024024 | 0.022999 | 0.025997 | NaN | NaN
  |   | SDC | 10000000 | 0.052 | 0.049 | 0.057 | 0.254009 | 0.000004
  | 4 | SDC | 10000000 | 0.033 | 0.021 | 0.037 | 0.297017 | 0.000998

Results on numbers.


name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.isin | 1 | Python | 10000000 | 0.368002 | 0.357003 | 0.390997 | NaN | NaN
  |   | SDC | 10000000 | 1.573 | 1.554 | 1.608 | 0.378001 | 0.375012
  | 4 | SDC | 10000000 | 1.079 | 1.074 | 1.092 | 0.339993 | 0.39201


Result on string

**Comment by Rubtsowa** [comment]
In perf tests 1 out of 2 tests is commented out. This is a temporary solution. If there is no comment, the first test will start anyway. If we want to look at the results of another, we will need to comment on the first test.

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.isin(values=[0]) | 1 | Python | 10000000 | 0.033634 | 0.028539 | 0.049437 |   |  
  |   | SDC | 10000000 | 0.148 | 0.11 | 0.203 | 0.371609 | 0.000861
Series.isin(values=[0]) | 4 | SDC | 10000000 | 0.034 | 0.033 | 0.046 | 0.280534 | 0


on numbers


name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.isin | 1 | Python | 10000000 | 0.251056 | 0.243026 | 0.282613 |   |  
SDC | 10000000 | 1.662 | 1.626 | 1.672 | 0.419142 | 0.360058
Series.isin | 4 | SDC | 10000000 | 1.275 | 1.246 | 1.429 | 0.622497 | 0.575027


on strings

**Comment by AlexanderKalistratov** [comment]
@densmirn what exactly do you mean?

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.isin | 1 | Python | 10000000 | 0.417631 | 0.392537 | 0.549333 |   |  
  |   | SDC | 10000000 | 1.477 | 1.43 | 1.614 | 0.370473 | 0.337034
Series.isin | 4 | SDC | 10000000 | 1.06 | 1.022 | 1.084 | 0.335106 | 0.348392

on strings

name | nthreads | type | size | median | min | max | compile | boxing
Series.isin(values=[0]) | 1 | Python | 10000000 | 0.022004 | 0.021565 | 0.027017 |   |  
  |   | SDC | 10000000 | 0.094 | 0.091 | 0.101 | 0.299716 | 5.75E-05
Series.isin(values=[0]) | 4 | SDC | 10000000 | 0.037 | 0.035 | 0.048 | 0.505522 | 3.36E-05

on numbers

**Comment by densmirn** [comment]
> @densmirn what exactly do you mean?

I meant [this](https://github.com/IntelPython/sdc/pull/603/commits/40b61b455dcce57766dca183edd6bf41d45f132e).

**Comment by AlexanderKalistratov** [comment]
@densmirn no. We don't want this. Why?

**Comment by densmirn** [comment]
@AlexanderKalistratov This lets us predictable scalability, doesn't?

**Comment by AlexanderKalistratov** [comment]
@densmirn no.

**Pull request merged**